src/EFCore/Metadata/Conventions/RelationshipDiscoveryConvention.cs

```
@@ -725,6 +783,90 @@ private IReadOnlyList<RelationshipCandidate> FindRelationshipCandidates(IConvent
+            var inverseName = inverse.GetSimpleMemberName();
+            if (((EntityType)leftEntityType)
+                    .FindNavigationsInHierarchy(navigationName).FirstOrDefault() != null
+                || ((EntityType)rightEntityType)
+                    .FindNavigationsInHierarchy(inverseName).FirstOrDefault() != null)
+            {
+                // navigations have already been configured using these properties
+                return;
+            }
+
+            leftSkipNavigation = leftEntityType.AddSkipNavigation(
+                navigationName, navigation, rightEntityType,
+                collection: true, onDependent: false, fromDataAnnotation: false);
+            if (leftSkipNavigation == null)
+            {
+                return;
+            }
+
+            rightSkipNavigation = rightEntityType.AddSkipNavigation(
+                inverseName, inverse, leftEntityType,
+                collection: true, onDependent: false, fromDataAnnotation: false);
+            if (rightSkipNavigation == null)
+            {
+                // Failed to create the right skip navigation - so remove
+                // the left skip navigation we just created as well.
+                leftEntityType.RemoveSkipNavigation(leftSkipNavigation);
+                return;
+            }
+
+            var associationEntityTypeBuilder =
+                ((InternalModelBuilder)entityTypeBuilder.ModelBuilder).AssociationEntity(
+                    (SkipNavigation)leftSkipNavigation,
+                    (SkipNavigation)rightSkipNavigation,
+                    ConfigurationSource.Convention);
+            if (associationEntityTypeBuilder == null)
+            {
+                // Failed to create the association entity type - so remove
+                // both the skip navigations we just created as well.
+                leftEntityType.RemoveSkipNavigation(leftSkipNavigation);
+                rightEntityType.RemoveSkipNavigation(rightSkipNavigation);
```

In what case would this actually remove the skip navigations?

---

Try debugging e.g. the test `Many_to_many_join_table_stored_in_snapshot`. In that case, this situation occurs when a 2nd entity type has been pulled in to the model because the 1st entity type references it, but `KeyDiscoveryConvention` has not yet run on the 2nd entity type - so it doesn't yet have a PK. Then the FK that uses that PK cannot be set up, `AssociationEntity()` returns `null` and we need to remove the skip navs.

Of course `KeyDiscoveryConvention` does run on 2nd entity type later, and this all gets resolved after that.

---

Ok, so we can remove this from here and let the association convention create it when it becomes possible

---

I don't think that works. This call knows that it specifically created the skip navigations and so it is safe and correct to remove them if the attempt to create the association entity fails. But `ManyToManyAssociationEntityTypeConvention` does not know that. It just knows that some skipNavs were created before it was called - it doesn't know how. I think the convention that creates the skipNavs should be responsible for removing them if necessary.

---

What I am saying is that we shouldn't try to remove the skip navigations. Just create the association entity type if possible and let it fail in model validation if not.